### PR TITLE
Sarah room and repo

### PR DESCRIPTION
**Pull request by SarahPanther**
Data-layer addition:

- BreatheRoomDatabase - our single database for the app; contains the Inhaler_Usage_Event_table and the Wearable_Data_table
- BreatheRepository - clean interface between app and data
- BreatheDAO - data access object for our database
- InhalerUsageEvent entity - represents a single, uniquely occuring Inhaler Usage Event 
- WearableData entity - represents a single, uniquely occurring cluster of data gathered from the smart pin

Included some simple tests.

### CODE REVIEW

File: Breathe/app/build.gradle [hunk 1]
```diff
@@ -20,15 +20,46 @@ android {
...
+    // Assertions
     androidTestImplementation 'androidx.test.ext:junit:1.1.2'
-    androidTestImplementation 'androidx.test.espresso:espresso-core:3.3.0'
+    androidTestImplementation 'androidx.test.ext:truth:1.3.0'
+    androidTestImplementation 'com.google.truth:truth:1.1.2'
+
+    // Android Architecture Component libraries
 
+    // Room components
+    implementation "android.arch.persistence.room:runtime:$rootProject.roomVersion"
```
Comment (Youssef-Beltagy) [inline Breathe/app/build.gradle]: Should this be androidx?

File: Breathe/.idea/inspectionProfiles/Project_Default.xml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<component name="InspectionProjectProfileManager">
```
Comment (Youssef-Beltagy) [inline Breathe/.idea/inspectionProfiles/Project_Default.xml:1]: What does this file do?

File: Breathe/app/src/main/java/com/ybeltagy/breathe/InhalerUsageEvent.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+ * <p>
+ * TODO: create static InhalerUsageEvent createIUE() to collect all data for InhalerUsageEvent and return an InhalerUsageEvent object
+ */
+@Entity(tableName = "InhalerUsageEvent_table")
+public class InhalerUsageEvent {
+    @PrimaryKey // timeStamp is the unique identifier for each InhalerUsageEvent record
+    @NonNull // this can never be null
+    // name of the column in Room that stores timeStamp
+    @ColumnInfo(name = "Inhaler_Usage_Event_UTC_ISO_8601_date_time")
+    private OffsetDateTime inhalerUsageEventTimeStamp;
```
Comment (Youssef-Beltagy) [inline Breathe/app/src/main/java/com/ybeltagy/breathe/InhalerUsageEvent.java]: Is it worthwhile to change this to instant?

File: Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    // IMPORTANT: this is here for testing use only; it should not be used in the app itself
+    // This deletes all InhalerUsageEvents from the InhalerUsageEvent_table
+    @Query("DELETE FROM InhalerUsageEvent_table")
+    void deleteAll();
+
+    // Returns all InhalerUsageEvents from the InhalerUsageEvent_table in lexographical
+    // ascending order of string timestamp (because of UTC format, this is in chronological order)
+    @Query("SELECT * FROM InhalerUsageEvent_table " +
+            "ORDER BY Inhaler_Usage_Event_UTC_ISO_8601_date_time ASC")
```
Comment (Youssef-Beltagy) [inline Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java]: Might Be better to get DESC

File: Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheRoomDatabase.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+@Database(entities = {InhalerUsageEvent.class, WearableData.class}, version = 1)
+@TypeConverters({Converters.class})
+public abstract class BreatheRoomDatabase extends RoomDatabase {
+
+    public abstract BreatheDao breatheDao();
+
+    private static volatile BreatheRoomDatabase INSTANCE; // this BreatheRoomDatabase is a singleton
+    private static final int NUMBER_OF_THREADS = 4;
+    // Executor Service is used (replaced AsyncTask) so inserts may happen concurrently
+    static final ExecutorService dbWriteExecutor = Executors.newFixedThreadPool(NUMBER_OF_THREADS);
```
Comment (Youssef-Beltagy) [inline Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheRoomDatabase.java]: Why use a fixed size thread pool?

Does it have any effect on a component's lifecycle?

File: Breathe/app/src/main/java/com/ybeltagy/breathe/WeatherData.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
+package com.ybeltagy.breathe;
+
+/**
+ * 3 state level (used for pollen index)
+ */
+enum Level {
```
Comment (Youssef-Beltagy) [inline Breathe/app/src/main/java/com/ybeltagy/breathe/WeatherData.java:6]: Can we rename this to PollenLevel?

File: Breathe/app/src/main/java/com/ybeltagy/breathe/InhalerUsageEvent.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    // Note: the java.time package and the OffsetDateTime class seem like the correct way to store
+    // our dates for Java 8
+    // (https://medium.com/decisionbrain/dates-time-in-modern-java-4ed9d5848a3e)
+
+    // I made these public because all of the Embedded examples I saw were public
+    // For more info on Embedded objects within Entities,
+    // see https://developer.android.com/training/data-storage/room/relationships
+    @Embedded public WeatherData weatherData;
```
Comment (Youssef-Beltagy) [inline Breathe/app/src/main/java/com/ybeltagy/breathe/InhalerUsageEvent.java]: You should be able to make it private as long as you make setters/getters.

File: .idea/Smart-Inhaler_Room.iml [hunk 8]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Youssef-Beltagy) [inline .idea/Smart-Inhaler_Room.iml]: You must delete this file.

File: Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    List<InhalerUsageEvent> getAllIUEs();
+
+    // Update one or more InhalerUsageEvents
+    @Update
+    void updateInhalerUsageEvent(InhalerUsageEvent... inhalerUsageEvents);
+
+
+    // Return InhalerUsageEvents between the parameter dates (inclusive)
+    @Query("SELECT * FROM InhalerUsageEvent_table " +
```
Comment (Youssef-Beltagy) [inline Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java:40]: String comparisons use length. So you need to make sure all the timestamps are the same lenght.

File: Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheRoomDatabase.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+ * - contains the Wearable_Data_table with WearableData entities
+ */
+@Database(entities = {InhalerUsageEvent.class, WearableData.class}, version = 1)
+@TypeConverters({Converters.class})
+public abstract class BreatheRoomDatabase extends RoomDatabase {
+
+    public abstract BreatheDao breatheDao();
+
+    private static volatile BreatheRoomDatabase INSTANCE; // this BreatheRoomDatabase is a singleton
+    private static final int NUMBER_OF_THREADS = 4;
```
Comment (Youssef-Beltagy) [inline Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheRoomDatabase.java]: Don't use fixed size thread pools.

File: Breathe/app/build.gradle (hunk 1 above)
Comment (SarahPanther) [reply Breathe/app/build.gradle]: Maybe, I will test replacing this with AndroidX when I work on LiveData.

File: Breathe/app/src/main/java/com/ybeltagy/breathe/InhalerUsageEvent.java (hunk 3 above)
Comment (SarahPanther) [reply Breathe/app/src/main/java/com/ybeltagy/breathe/InhalerUsageEvent.java]: Yes, thank you for changing this to instant.

File: Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java (hunk 4 above)
Comment (SarahPanther) [reply Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java]: Changed to DESC

File: Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheRoomDatabase.java (hunk 5 above)
Comment (SarahPanther) [reply Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheRoomDatabase.java]: changed to cachedThreadPool

File: Breathe/app/src/main/java/com/ybeltagy/breathe/InhalerUsageEvent.java (hunk 7 above)
Comment (SarahPanther) [reply Breathe/app/src/main/java/com/ybeltagy/breathe/InhalerUsageEvent.java]: Changed to private

File: .idea/Smart-Inhaler_Room.iml (hunk 8 above)
Comment (SarahPanther) [reply .idea/Smart-Inhaler_Room.iml]: Deleted

File: Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheRoomDatabase.java (hunk 10 above)
Comment (SarahPanther) [reply Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheRoomDatabase.java]: Changed to cachedThreadPool

**Comment by SarahPanther** [comment]
- Left unresolved comments so we can keep them in mind as we move forward
- Resolved comments on things Youssef fixed in his review branch of this branch

**Pull request merged**